resources/lib/library.py

```diff
@@ -1305,6 +1405,122 @@ def addons( self ):
+        try:
+            tree = xmltree.parse( os.path.join( path, "addon.xml" ) ).getroot()
+            for extension in tree.findall( "extension" ):
+                if "point" in extension.attrib and extension.attrib.get( "point" ) == "xbmc.python.pluginsource":
+                    # Find out what content type it provides
+                    provides = extension.find( "provides" )
+                    if provides is None:
+                        return None
+                    return provides.text.split( " " )
+
+
+        except:
+            return None
+        return None
+    
+    def detectPluginContent(self, item):
+        #based on the properties in the listitem we try to detect the content
+        
+        if not item.has_key("showtitle") and not item.has_key("artist"):
+            #these properties are only returned in the json response if we're looking at actual file content...
+            # if it's missing it means this is a main directory listing and no need to scan the underlying listitems.
+            return None
+
+        if not item.has_key("showtitle") and item.has_key("artist"):
+            ##### AUDIO ITEMS ####
+            if item["artist"][0] == item["title"]:
+                return "artists"
+            elif item["album"] == item["title"]:
+                return "albums"
+            elif (item["type"] == "song" or (item["artist"] and item["album"])):
+                return "songs"
+        else:    
+            ##### VIDEO ITEMS ####
+            if (item["showtitle"] and not item["artist"]):
+                #this is a tvshow, episode or season...
+                if (item["season"] > -1 and item["episode"] == -1):
+                    return "seasons"
+                elif item["season"] > -1 and item["episode"] > -1:
+                    return "episodes"
+                else:
+                    return "tvshows"
+            elif (item["artist"]):
+                #this is a musicvideo!
+                return "musicvideos"
+            elif (item["imdbnumber"] or item["mpaa"] or item["trailer"] or item["studio"]):
+                return "movies"
+
+        return None
+
+    def widgets( self ):
+        if self.loadedWidgets == True:
+            # The list has already been populated
+            return self.loadedWidgets
+        elif self.loadedWidgets == "Loading":
+            # The list is currently being populated, wait and then return it
+            count = 0
+            while count < 20:
+                xbmc.sleep( 100 )
+                count += 1
+                if self.loadedWidgets == True:
+                    return self.loadedWidgets
+        else:
+            # We're going to populat the list
+            self.loadedWidgets = "Loading"
+
+        # Load skin overrides
+        tree = DATA._get_overrides_skin()
+
+        # Get widgets
+        listitems = []
+        log( "Loading skin widgets" )
+        try:
```

Comment:
He he, yup I'll look again at the widget() function (and pretty much all the library loading functions, which I copied/pasted and stuck the widget code into!)